receiver/snmpreceiver/client.go

```diff
@@ -0,0 +1,512 @@
+		goSNMP.SetTransport(lCaseScheme)
+	}
+
+	// Checked in config that it exists
+	portStr := snmpURL.Port()
+
+	port, err := strconv.ParseUint(portStr, 10, 16)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create goSNMP client: issue parsing port '%s'. %w", portStr, err)
+	}
+	goSNMP.SetPort(uint16(port))
+
+	// Set goSNMP target based on config
+	goSNMP.SetTarget(snmpURL.Hostname())
+
+	if goSNMP.GetVersion() == gosnmp.Version3 {
+		// Set goSNMP v3 configs
+		setV3ClientConfigs(goSNMP, cfg)
+	} else {
+		// Set goSNMP community string
+		goSNMP.SetCommunity(cfg.Community)
+	}
+
+	// return client
+	return &snmpClient{
+		client: goSNMP,
+		logger: logger,
+	}, nil
+}
+
+// setV3ClientConfigs sets SNMP v3 related configurations on gosnmp client based on config
+func setV3ClientConfigs(client goSNMPWrapper, cfg *Config) {
+	client.SetSecurityModel(gosnmp.UserSecurityModel)
+	// Set goSNMP user based on config
+	securityParams := &gosnmp.UsmSecurityParameters{
+		UserName: cfg.User,
+	}
+	// Set goSNMP security level & auth/privacy details based on config
+	switch strings.ToUpper(cfg.SecurityLevel) {
+	case "NO_AUTH_NO_PRIV":
+		client.SetMsgFlags(gosnmp.NoAuthNoPriv)
+	case "AUTH_NO_PRIV":
+		client.SetMsgFlags(gosnmp.AuthNoPriv)
+		protocol := getAuthProtocol(cfg.AuthType)
+		securityParams.AuthenticationProtocol = protocol
+		securityParams.AuthenticationPassphrase = cfg.AuthPassword
+	case "AUTH_PRIV":
+		client.SetMsgFlags(gosnmp.AuthPriv)
+
+		authProtocol := getAuthProtocol(cfg.AuthType)
+		securityParams.AuthenticationProtocol = authProtocol
+		securityParams.AuthenticationPassphrase = cfg.AuthPassword
+
+		privProtocol := getPrivacyProtocol(cfg.PrivacyType)
+		securityParams.PrivacyProtocol = privProtocol
+		securityParams.PrivacyPassphrase = cfg.PrivacyPassword
+	}
+	client.SetSecurityParameters(securityParams)
+}
+
+// getAuthProtocol gets gosnmp auth protocol based on config auth type
+func getAuthProtocol(authType string) gosnmp.SnmpV3AuthProtocol {
+	var authProtocol gosnmp.SnmpV3AuthProtocol
+
+	switch strings.ToUpper(authType) {
+	case "MD5":
+		authProtocol = gosnmp.MD5
+	case "SHA":
+		authProtocol = gosnmp.SHA
+	case "SHA224":
+		authProtocol = gosnmp.SHA224
+	case "SHA256":
+		authProtocol = gosnmp.SHA256
+	case "SHA384":
+		authProtocol = gosnmp.SHA384
+	case "SHA512":
+		authProtocol = gosnmp.SHA512
+	}
+
+	return authProtocol
+}
+
+// getPrivacyProtocol gets gosnmp privacy protocol based on config privacy type
+func getPrivacyProtocol(privacyType string) gosnmp.SnmpV3PrivProtocol {
+	var privacyProtocol gosnmp.SnmpV3PrivProtocol
+
+	switch strings.ToUpper(privacyType) {
+	case "DES":
+		privacyProtocol = gosnmp.DES
+	case "AES":
+		privacyProtocol = gosnmp.AES
+	case "AES192":
+		privacyProtocol = gosnmp.AES192
+	case "AES192C":
+		privacyProtocol = gosnmp.AES192C
+	case "AES256":
+		privacyProtocol = gosnmp.AES256
+	case "AES256C":
+		privacyProtocol = gosnmp.AES256C
+	}
```

Comment:
Same. Use a default or check return value.

Reply:
@djaglowski So this is all validated in the client. so this "shouldn't' be possible. I'll switch to having a default case anyways.